cmake/modules/RootMacros.cmake

```
@@ -1232,35 +1234,77 @@ function(ROOT_INSTALL_HEADERS)
     install(DIRECTORY ${d} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
                            COMPONENT headers
                            ${options})
     string(REGEX REPLACE "(.*)/$" "\\1" d ${d})
-    ROOT_GLOB_FILES(include_files
+    ROOT_GLOB_FILES(globbed_files
       RECURSE
-      RELATIVE ${CMAKE_CURRENT_SOURCE_DIR}/${d}
+      RELATIVE ${CMAKE_CURRENT_SOURCE_DIR}
       FILTER ${filter}
       ${d}/*.h ${d}/*.hxx ${d}/*.icc )
-    foreach (include_file ${include_files})
-      set (src ${CMAKE_CURRENT_SOURCE_DIR}/${d}/${include_file})
-      set (dst ${CMAKE_BINARY_DIR}/include/${include_file})
-      add_custom_command(
-        OUTPUT ${dst}
-        COMMAND ${CMAKE_COMMAND} -E copy ${src} ${dst}
-        COMMENT "Copying header ${src} to ${CMAKE_BINARY_DIR}/include"
-        DEPENDS ${src})
-      list(APPEND dst_list ${dst})
-    endforeach()
+    list(APPEND include_files ${globbed_files})
+  endforeach()
+
+  string(REPLACE ${CMAKE_SOURCE_DIR} "" target_name ${CMAKE_CURRENT_SOURCE_DIR})
+  string(REPLACE / _ target_name "copy_header_${target_name}")
+  string(REGEX REPLACE "_$" "" target_name ${target_name})
+
+  # Register the files to be copied for each target directory (e.g. include/ include/ROOT include/v7/inc/ ...)
+  list(REMOVE_DUPLICATES include_files)
+  list(TRANSFORM include_files REPLACE "(.*)/[^/]*" "\\1/" OUTPUT_VARIABLE subdirs)
+  list(REMOVE_DUPLICATES subdirs)
+  foreach(subdir ${subdirs})
+    set(input_files ${include_files})
+    list(FILTER input_files INCLUDE REGEX "^${subdir}[^/]*$")
+    set(output_files ${input_files})
+
+    string(REGEX REPLACE ".*/*inc/" "" destination ${subdir})
+
+    list(TRANSFORM input_files  PREPEND "${CMAKE_CURRENT_SOURCE_DIR}/")
+    list(TRANSFORM output_files REPLACE ".*/" "${CMAKE_BINARY_DIR}/include/${destination}")
+
+    set(destination destination_${destination})
+
+    set_property(GLOBAL APPEND PROPERTY ROOT_HEADER_COPY_LISTS ${destination})
+    set_property(GLOBAL APPEND PROPERTY ROOT_HEADER_INPUT_${destination} ${input_files})
+    set_property(GLOBAL APPEND PROPERTY ROOT_HEADER_OUTPUT_${destination} ${output_files})
   endforeach()
-  if (dst_list)
-    string(REPLACE ${CMAKE_SOURCE_DIR} "" tgt ${CMAKE_CURRENT_SOURCE_DIR})
-    string(MAKE_C_IDENTIFIER move_header${tgt} tgt)
-    set_property(GLOBAL APPEND PROPERTY ROOT_HEADER_TARGETS ${tgt})
-    add_custom_target(${tgt} DEPENDS ${dst_list})
-  endif()
 endfunction()
 
+#---------------------------------------------------------------------------------------------------
+#--- ROOT_CREATE_HEADER_COPY_TARGETS
+#    Creates a target to copy all headers that have been registered for copy in ROOT_INSTALL_HEADERS
+#---------------------------------------------------------------------------------------------------
+macro(ROOT_CREATE_HEADER_COPY_TARGETS)
+  get_property(HEADER_COPY_LISTS GLOBAL PROPERTY ROOT_HEADER_COPY_LISTS)
+  list(REMOVE_DUPLICATES HEADER_COPY_LISTS)
+  foreach(COPY_LIST ${HEADER_COPY_LISTS})
```

```suggestion
  foreach(copy_list ${HEADER_COPY_LISTS})
```
We seems to tend to use lower case for variables ...

---

OK ...